cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+    ValueType query_cell( int cellid_x, int cellid_y, int cellid_z )
+    {
+        auto tileid = _blkIdSpace.find( cellid_x >> TileBitsPerDim,
+                                        cellid_y >> TileBitsPerDim,
+                                        cellid_z >> TileBitsPerDim );
+        auto cellid = _tileIdSpace.coord_to_offset(
+            cellid_x & TileMask, cellid_y & TileMask, cellid_z & TileMask );
+        return ( ValueType )( ( tileid << TileBits ) & cellid );
+    }
+
+  private:
+    //! block index space, map tile ijk to tile No
+    BlockIndexSpace<TileBitsPerDim, TileSizePerDim, TSize, HashType, Device,
+                    KeyType, ValueType>
+        _blkIdSpace;
+    //! tile index space, map cell ijk to cell local No inside a tile
+    TileIndexSpace<TileBitsPerDim, TileSizePerDim, TSize> _tileIdSpace;
+    //! space size (global), channel size
+    Kokkos::Array<int, Rank> _min;
+    Kokkos::Array<int, Rank> _max;
+};
+
+//---------------------------------------------------------------------------//
+template <unsigned long long TBits, unsigned long long TSizePerDim,
+          unsigned long long TSize, typename Hash, typename Device,
+          typename Key, typename Value>
+class BlockIndexSpace
+{
+  public:
+    //! Number of cells inside each tile
+    static constexpr unsigned long long TileBitsPerDim = TBits;
+    static constexpr unsigned long long TileSizePerDim = TSizePerDim;
+    static constexpr unsigned long long TileSize = TSize;
+    //! Types
+    using KeyType = Key;     // tile hash key type
+    using ValueType = Value; // tile value type
+    using HashType = Hash;   // hash table type
+
+    BlockIndexSpace( int capacity, float rehash_factor,
+                     std::array<int, N> &size )
+        : _tile_num( 0 )
+        , _tile_capacity( capacity )
+        , _rehash_coeff( rehash_factor )
+        , _tile_table( capacity )
+        , _op_ijk2key( size[0], size[1], size[2] )
+        , _op_key2ijk( size[0], size[1], size[2] )
+    {
+        _block_table.clear();
+        _value_table.clear();
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    ValueType insert( int tileid_x, int tileid_y, int tileid_z )
+    {
+        if ( _tile_table.size() >= _tile_capacity )
+        {
+            _tile_capacity = static_cast<int>( _tile_capacity * _rehash_coeff );
+            _tile_table.rehash( _tile_capacity );
+            // TODO: view reallocate needed
+        }
+        const KeyType tileKey = _op_ijk2key( tileid_x, tileid_y, tileid_z );
+        ValueType tileNo;
+        if ( ( tileNo = _key_table.find( tileKey ) ) ==
+             Kokkos::UnorderedMap::invalid_index )
+        {
+            // current impl: use inserting sequence as id
+            // rethink: sort the id .
+            tileNo = _blk_table.size();
+            _tile_table.insert( tileKey, tileNo );
+        }
+        return tileNo;
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    void sortTile()
+    {
+        // pass
```

Comment:
would your objective be to sort the map here?

Reply:
Yes, I'm trying to see if we can sort after all tiles are inserted. Since each tile is mapped to a Morton code, what we should do here is to get all the Morton code inside the hash table, and sort them, and then give them a new id.